### DISCUSSION TITLE
Fake part with Fake timings

### DISCUSSION THREAD

**Discussion by bartokon** [JulianKemmerer/PipelineC#42] [Ideas]
Hi,
I'm thinking if it is possible to create a fake part with fake timings, so whole place and route could take seconds (Cached synthesis p&r?) That part should have unlimited resources and each path delay could be for example 1.5*logic_delay. Logic delay could be avg from different FPGA families. All latency finding algorithms could much faster to iterate...

For example generate netlist from Pipeline parsed code then generate fake timing paths depending on logic depth and wire length.

**Comment by JulianKemmerer** [comment]
`so whole place and route could take seconds` depends on what you mean?
In theory yes - but depending on how much of GHDL->Yosys->nextpnr I want to `fake` it could be a rather large undertaking to do so.

I mentioned in another issue there is two slow syn+pnr processes
1. Get comb. logic delay for all modules (so I know where to put pipeline regs)
2. Run syn+pnr with guess pipeline params based on where I think they should go (refine guess, repeat)

Right now its plausible to reduce the time spent doing 1) (the smaller of the two chunks since not iterative, and highly parallelizeable, but still quite slow) by not synthesizing all functions, but instead only synthesizing the most primitive operations (i.e. in cache) and lego-bricking building up a `fake` guess of how much delay larger user C functions have (quickly, as opposed to running full syn+pnr on all user code).

For 2) though - building up a fake map of pipelined params -> timing delays ... thats like creating a whole other tool , practically my own syn+pnr flow idk :woozy_face:

**Comment by bartokon** [comment]
Yes, you are right. 

For syn+pnr, I'm thinking about TerosHDL generating generic schematic with help of Yosys, so it gives us netlist to analyze. 
Now we could find an amount of logic levels to next register and assume that regardless of what wire or reg delay is always some 1 unit for example 1ns and passing reg is 1.5 unit and sum the worst path delay ang logic delay, put registers there. We don't even need to place and route, just use schematic to estimate performance. 

Idk I'm looking at: 
![image](https://user-images.githubusercontent.com/47795888/141082268-a4f8c530-90a3-42f8-8b9c-53203181d76e.png)
and I see that there is correlation in implementation, so few fake synth runs, and you could predict what latency gives you what frequency.
Just use hard-coded values for Virtex/Artix family idk average path delay in some basic implementation and use it as base unit.

**Reply by JulianKemmerer** [reply]
You might be interested in things like
https://antmicro.com/blog/2021/09/symbiflow-fpga-interchange-format/

**Reply by JulianKemmerer** [reply]
`amount of logic levels to next register ` does yosys provide this? Or do you need nextpnr for that? yosys also has a  `-ltp` option IIRC that is the longest topological path which is sorta of what you are getting at perhaps?

I worry though that this kind of very coarse grain roughly faked approach will not be enough to produce well pipelined designs after going through the maze of syn+pnr, device specific optimizations, etc kind of stuff. I started PipelineC with looser guesses based on the timing of primitives sort of like I mentioned before - and it wasnt accurate enough to make pipelines that met timing like I expected on actual FPGA hardware

**Reply by JulianKemmerer** [reply]
`sum the worst path delay` traversing a schematic like this is something closer to what nextpnr could do today than PipelineC could do _at some point_. I worry even in trying to make a faked faster version, idk being in python to start, it might not even be faster than nextpnr? :-/

**Reply by bartokon** [reply]
Hi Julian, was thinking about this:
http://www.clifford.at/yosys/cmd_write_json.html + write simple app to detect logic levels from json.
If you could add this write_json to output as well as CSV I could try to write logic level detector. 


PS. I have ran ~512 sequential runs on Vivado and Yosys. Yosys is faster to like ~30 latency and then Vivado is much faster in synth.

PS2. Who knows maybe Yosys already have some option to write maximum comb logic levels between regs?

**Reply by JulianKemmerer** [reply]
@bartokon OK we are part of the way there - the flow that utilizes nextpnr already uses the json output from yosys. So running the PipelineC tool today (set to use GHDL->Yosys->nextpnr flow via selecting ECP5 Lattice part) will produce that file for you now.

I imagine you want some kind of switch that says 'stop after writing the json from Yosys, dont run nextpnr' ?

And then is the CSV the full, each instance and params listed (not just latency, fmax) that you want now?

Refresh me as to what you have planned:

You want to run PipelineC->VHDL->GHDL->Yosys->Max logic levels between regs

And with that LLs count you plan to do what? Count each LL as 1 time unit and derive some fmax estimate? (saving time over taking json through nextpnr flow is the idea)

**Reply by bartokon** [reply]
Nextpnr is ultra slow compared to Vivado for bigger designs. 

So what is the goal?
For example, I'm using PipelineC to generate some optimal function.
I use pipelineC, with some part for example "ECP5" or custom "detect" part that also that doesn't run nextpnr (idk what is simpler to detect pragma part or add another arg to pipelineC command?). I will implement "logic" detector that will list logic levels with no clock in between to detect "slowest" function, for example logic stages to clock in fun1 = 5, fun2 = 4, fun3 = 5. So now I could bet that fun1 and fun3 will be the greatest bottleneck in design.  So I force pipelineC by CSV file to slice more fun1 and fun3 to 4 logic stages. This should pump up frequency to the next stage. 
I need to upload here a graph from mypipeline implementation. It goes like stairs, there are +- 4 freq stages that mean something to design performance and the rest from all this 512 iterations is just loosing time in implementation without any gain.

If you know there is an nextpnr flow that estaminates that latency quicker than +20mins that would be great :)

TL;DR as you have said, derive some fmax estimate much broader and much faster than place and route.

If Vitis HLS can estimate latency and performance in 1 second, why can't we?

After some thinking the easiest would be to add arg for force use Yosys to generate this json regardless of used part.

**Reply by JulianKemmerer** [reply]
OK what you are saying is sounding good to me

`If Vitis HLS can estimate latency and performance in 1 second, why can't we?` :sunglasses: _Let's do it then_

I agree some option to 'give me the yosys json' makes sense to have, will look into that

**Reply by JulianKemmerer** [reply]
Hmm @bartokon hit a bump?
Idk why by yosys doesnt have a json output when you dont specify the FPGA part?
Ex. see `-json` arg here
http://webcache.googleusercontent.com/search?q=cache:lLCoLZNOQhQJ:www.clifford.at/yosys/cmd_synth_ice40.html+&cd=6&hl=en&ct=clnk&gl=us


But now here for regular (no part specified) `synth` 
http://webcache.googleusercontent.com/search?q=cache:-ednjy8-SJEJ:www.clifford.at/yosys/cmd_synth.html+&cd=1&hl=en&ct=clnk&gl=us

```
ERROR: Command syntax error: Unknown option or option in arguments.
> synth -top top_8b62 -json top_8b62.json
```

Hmmm

**Reply by JulianKemmerer** [reply]
I think write_json command will work

**Reply by JulianKemmerer** [reply]
Ok sorry for the multiple pings @bartokon but, pull repo, you can now run 
`./src/pipelinec ./examples/pipeline.c --yosys_json`

And it ends getting us to a point where we can start moving forward I think:
```
================== Beginning Throughput Sweep ================================
Function: my_pipeline Target MHz: 250.0
Setting all instances to comb. logic to start...
Running Yosys on unpipelined top level comb. logic. TODO: Complete feedback for pipeline params...
Running synthesis on comb. logic only top level ...
Running: /home/julian/pipelinec_output/top/top_8b62.sh
Stopping after json output in: /home/julian/pipelinec_output/top
```
The state is:
I have this comb. logic `my_pipeline` I want to pipeline - but first I need to send the json to your tool so you can say back to me its 'N levels of logic and you should put a register in this func3 _here_'? Am I thinking right?
[top_8b62.json.txt](https://github.com/JulianKemmerer/PipelineC/files/7534286/top_8b62.json.txt)

**Reply by bartokon** [reply]
I'm looking at this json file and...
![image](https://user-images.githubusercontent.com/47795888/141695168-fc8eb051-50c6-45bf-83b0-fd1a0681559d.png)
I could analyze bin_op, mux_uint, my_pipeline and top. Go thought all the cells and find where the "distance" between non clocked cells is the greatest and return that cell instances and max latency? 
IDK what I could slice with pipelineC CSV parameters, but If I assume with your API I could tell slice bin_op_gt_uint1 more and retry analysis.

**Reply by bartokon** [reply]
Maybe you could take that list of cells from me and insert pipeline stages there? That would be sniper shots and guessing game would be not needed. What do you think? 

After a quick shower, let's not overengineer this simple ML find optimal structure idea. Just give me an option to generate json file with -start -stop and this will be great!

**Reply by JulianKemmerer** [reply]
Hmmm I am not sure how we want to proceed...

Can you tell me, from the json, how many logic levels each cell has? (ex making up. mult is 20 LLs, but add is just 5 idk)

I need that since from there, when you tell me '---start N' I know when those N slices go. I can't start at N unless I know where the N slices go.

This gets to the two phases of information I need faked instead of running nextpnr:

1. Get comb. logic delay (we can use logic levels as analogous) for all modules (so I know where to put pipeline regs)
2. Run syn+pnr with guess pipeline params based on where I think they should go, ex. --start N (refine guess, repeat)

I need the first info before beginning the --start N second part?

**Reply by bartokon** [reply]
ATM, I could give how many logic levels have one module, by following connection nets and finding the longest path to clock.
The best I could do is to give you max latency NETS list.

![image](https://user-images.githubusercontent.com/47795888/141701017-95517be7-6daf-42ad-ba93-999e9c1e90b6.png)

Input would be for example NET 2. 
Algorithm search for module where NET 2 is an input. 
Then look at output NET 194. 
Marks that NET 2 is visited if no clock detected increment latency, if max_latency is greater than atm_latency override max_latency else clock detected set atm_latency to 0.
Call itself for NET 194 as input.
Break loop if all NETS are visited.

**Reply by JulianKemmerer** [reply]
Hmmm I think thats close

Not sure what you mean by `how many logic levels have one module` ?

What you said sounds reasonable. Tracing input to outputs, accumulating some max delay/number of logic levels.

It looks like you would be accumulating up Yosys primitives named like `_ANDNOT_`
```json
"$abc$52550$auto$blifparse.cc:381:parse_blif$52563": {
          "hide_name": 1,
          "type": "$_ANDNOT_",
          "parameters": {
          },
          "attributes": {
          },
          "port_directions": {
            "A": "input",
            "B": "input",
            "Y": "output"
          },
          "connections": {
            "A": [ 137 ],
            "B": [ 142 ],
            "Y": [ 143 ]
          }
        },
```

And I would want to know, ex. making it up, in module `bin_op_gt...` the longest chain of primitives from that modules inputs ot outputs is a path through 

`3 ANDNOT, 4 ORNOT, and a NAND`. (just path through that module from input to output ports - not from reg to reg - this is comb logic, no regs yet).

And then from there we need to allocate some 'time/delay' unit to each of those (ex. ANDNOT is 1 unit, NAND is 2 units idk)
, and accumulate that.

Then we could say ex.  `bin_op_gt...`  is estimated to have 3x the delay of some other func - etc. I.e knowing where to add pipelining

**Reply by JulianKemmerer** [reply]
(if you _need_ a path between registers, we can run Yosys on each module individually (where I can put IO regs), instead of looking at json for the entire top level)

**Reply by bartokon** [reply]
I would search for connections and if type = _DFF_ (Dflipflop? I think it is only clocked element) then it is registered do max_logic_delay = 0.
I will make a function that takes f(module, net) and returns longest net in module. If no regs on path detected then logic levels = delays. If reg detected logic_delay = 0. Something like that.
I think that "net" is global for all modules so best is you could "run Yosys on each module individually". This should make this process faster because we don't need to parse whole modules just one module with its nets. Just put these jsons to some folder. 
Idk but now we could use json(Yosys parse data) to json(module slice params) as pipelineC pair for synthesis and  slicing? 
Maybe in format somemodule.json somemodule_params.json?

**Reply by JulianKemmerer** [reply]
I think you have the general idea.

Maybe we can talk about the expected results for each step if this works as expected on the `my_pipeline` example design.

We need to define `length/delay` for primitives, made up example. `ANDNOT is 1 unit, NAND is 2` .
Then using that length measure, traverse the netlist looking for the longest DFF->DFF path.

Then PipelineC needs to interrogate that path - what primitives are chained in sequence - so when asked to 'cut than into N stages' I know where to put new registers

Ill get back about the `my_pipeline` expected results walk through (maybe we can choose an even simpler example)

**Reply by bartokon** [reply]
Right now, I'm testing this follow net idea

I just need to change net list path detection, because it just adds parallel path as serial because of   list_of_path_sets.append(actual_path_set).

If you could create less difficult JSON that I could use as testbench that would be great.

**Reply by bartokon** [reply]
[test.txt](https://github.com/JulianKemmerer/PipelineC/files/7541701/test.txt)
Ahh, now I have figured out this upload function.

Notes:
1. I need to make it actual_path independent, because if DFF is detected, actual_path should be zeroed.

**Reply by JulianKemmerer** [reply]
I will try to put together a smaller example this evening :+1:

**Reply by JulianKemmerer** [reply]
OK here is the json for this code:
```c
// A II=1 pipeline from a pure C function (no globals or local static vars)

// Set FPGA part/synthesis tool see ../main.c for examples.
//#pragma PART "xc7a35ticsg324-1l"

#include "intN_t.h"
#include "uintN_t.h"

#pragma MAIN_MHZ my_pipeline 250.0
uint8_t my_pipeline(uint8_t x)
{
  uint8_t rv = x;
  if(x == 0)
  {
    rv = !x;
  }
  return rv;
}
```
[top_019d.json.txt](https://github.com/JulianKemmerer/PipelineC/files/7542514/top_019d.json.txt)

**Reply by JulianKemmerer** [reply]
And for 
```c
// A II=1 pipeline from a pure C function (no globals or local static vars)

// Set FPGA part/synthesis tool see ../main.c for examples.
//#pragma PART "xc7a35ticsg324-1l"

#include "intN_t.h"
#include "uintN_t.h"

#pragma MAIN_MHZ my_pipeline 250.0
uint8_t my_pipeline(uint8_t x)
{
  return x + 1;
}
```
[top_265f.json.txt](https://github.com/JulianKemmerer/PipelineC/files/7542558/top_265f.json.txt)

**Reply by bartokon** [reply]
Uint8 is like 8 full adders + registers, on gate level. ~1k lines of json for me to search... I need something REALLY simple to just verify if netlist path detector works.

Could you also generate? 
`// A II=1 pipeline from a pure C function (no globals or local static vars)

// Set FPGA part/synthesis tool see ../main.c for examples.
//#pragma PART "xc7a35ticsg324-1l"

#include "intN_t.h"
#include "uintN_t.h"

#pragma MAIN_MHZ my_pipeline 250.0
bool my_pipeline(uint8_t x)
{
  return !x;
}`
Or even something in plain vhdl?

**Reply by JulianKemmerer** [reply]
The plain VHDL is pretty easy to read - are you thinking to try and use that as some kind of netlist?

I dont really have control on how the netlist looks after the GHDL->Yosys flow is done with it hmm...

Let me see if I can remove the top level wrapper around `my_pipeline` and try your `return !x;` example

**Reply by JulianKemmerer** [reply]
OK, pull repo
Now if you run with `--coarse --yosys_json` it will see you only have a single `MAIN` function and coarse write a hopefully smaller json file of just your single function and no top level wrapper around that as is default.

```
// A II=1 pipeline from a pure C function (no globals or local static vars)

// Set FPGA part/synthesis tool see ../main.c for examples.
//#pragma PART "xc7a35ticsg324-1l"

#include "intN_t.h"
#include "uintN_t.h"

#pragma MAIN_MHZ my_pipeline 250.0
uint8_t my_pipeline(uint8_t x)
{
  return !x;
}
```
[my_pipeline_0CLK_b45f1687_top.json.txt](https://github.com/JulianKemmerer/PipelineC/files/7548577/my_pipeline_0CLK_b45f1687_top.json.txt)

(Note this .json now does have register/flip flops, its input regs -> not op -> output regs - previous .json files were pure combinatorial logic)

**Reply by bartokon** [reply]
Julian, sorry for spamming but look at this!

We can use Yosys to generate blif file by using write_blif command (This should be equivalent to JSON in time needed).
Then we can pass that blif to PYRTL, see example 7 and 8 to extract worst timing paths and insert regs there!
https://github.com/UCSBarchlab/PyRTL/tree/cf5a682313f20be204ed6ea586ce51ce6861c998/ipynb-examples

This should be 300% easier than writing all from scratch

**Reply by bartokon** [reply]
It is working! just change write_json to write_blif
![image](https://user-images.githubusercontent.com/47795888/142077765-e87a25df-1c83-4519-82a8-742f9eb67c84.png)
[my_pipeline_0CLK_6f2c5aad_top.blif.txt](https://github.com/JulianKemmerer/PipelineC/files/7550353/my_pipeline_0CLK_6f2c5aad_top.blif.txt)

https://pyrtl.readthedocs.io/en/latest/analysis.html#estimation
we can even get estimated frequency:
timing.max_freq(tech_in_nm=65)

PS. Can you add --yosys_blif with --start n --stop n working? :D 
I must test if this method is faster than synthesis if latency is like 200 cycles.

**Reply by JulianKemmerer** [reply]
Oh wow that pyrtl estimation functionality sounds very useful indeed! If its really that simple in python maybe the PipelineC tool can drop in pyrtl based estimates as if it were its own synthesis tool - all I need is `total block timing delay is` number (and eventually maybe having better net/modules names ex. what is `tmp465`?)

Maybe we dont even need `--yosys_blif` if use of pyrtl is built in

(And btw the `--start` and `--stop` will need to come after we get this first phase of feedback/knowledge of the delay of things)

**Reply by JulianKemmerer** [reply]
Yeah I am feeling pretty certain that being able to select `pyrtl` as your `fake` synthesis tool + timing numbers source is good idea

**Reply by bartokon** [reply]
If top main is just one function, then we can use this as pyrtl basic block to get block delay
![image](https://user-images.githubusercontent.com/47795888/142155824-5a549e7b-054e-421c-9fd9-4b399208e8ef.png)
We just need to pass each function to yosys and get blif for it

**Reply by JulianKemmerer** [reply]
Exactly that is part 1 of the timing optimization. Every function is synthesized (here run through pyrtl estimations) as a top level of a single function to get the delay per function.

Then multiple functions are combined in a pipelined top level and the iterative changes begin
This time the iteration should be alot faster using pyrtl 

I am planning to start as pyrtl being like another stand alone syn tool (like other syn tools I support)

And then slowly work in integrating pyrtl-the-python-library as the default fast estimator even for when using other tools (estimate with pyrtl, final syn with vivado or something).

**Reply by bartokon** [reply]
So basically you will add pyrtl support to pipelinec and accelerate functionality many times 🗡️

I'm waiting for some sneak peek on your work!

Btw. Don't forget about more knobs :3

**Reply by JulianKemmerer** [reply]
Precisely about more knobs - especially for that last step I mentioned `slowly work in integrating pyrtl-the-python-library as the default fast estimator even for when using other tools (estimate with pyrtl, final syn with vivado or something)`

There will need to be some tuning of how the pyrtl estimator maps to vivado results
So when we are doing doing our `many times` accelerated estimation stuff - it is getting towards the goal for vivado too (it should be in the ballpark to start I think)

**Reply by JulianKemmerer** [reply]
And im sure you want more pipelinec knobs too ;)

**Reply by JulianKemmerer** [reply]
You wanted to see it as soon as it happened @bartokon 
Just pushed what looks like working support for basic pyrtl use

Comment out your `#pragma PART` line and if you have pyrtl installed it will default to it
And I tested with `--coarse` since its a single pipeline example we've been using

**Reply by JulianKemmerer** [reply]
Will be trying bigger examples this evening hopefully

**Comment by bartokon** [comment]
Okay Julian, I will research this myself and if I find something usable I will write about it here :)

PS. Are use using incremental synthesis or similar flow that could speedup synth process?

**Reply by JulianKemmerer** [reply]
Lovely - I await hear on what you find in your research :nerd_face: 

I am not using any flows like you are thinking I don't believe. Right now each module that is synthesized is simply its own top level, run independently in an out-of-context mode synthesis (talking Vivado here). I havent done partial/incremental synthesis since I dont think designs are large enough / partitioned well enough (i.e. into separate large chunks) for that be beneficial most of the time

**Reply by bartokon** [reply]
Look at this:
https://support.xilinx.com/s/article/976545?language=en_US
adding reg and re-synth 50% synth time. 
It should find target latency faster for at least Vivado :)

**Reply by bartokon** [reply]
I will try to test synthesis model reuse by using open tools to create base model and then try to move it to Vivado :p

**Reply by JulianKemmerer** [reply]
Hmm interesting - ive never seen incremental synthesis used.

It is not very frequent that I even have the `5 RTL Partitions of size greater than 10K instances` size requirement to trigger parallel synthesis.

But yeah once reaching that size this does seem like it would speed things up. Good to know about!

**Comment by JulianKemmerer** [comment]
I will get to your other comments @bartokon but btw
I really appreciate these constructive conversions we are having regarding improving PipelineC

I am often on this Discord's #hdl-other channel talking about PipelineC
https://discord.com/invite/pY2wJvnCMY
If you are into that sort of thing - stop on by, spark up some conversation there :)

**Comment by JulianKemmerer** [comment]
See issue https://github.com/JulianKemmerer/PipelineC/issues/46
for further work+discussion

**Comment by JulianKemmerer** [comment]
Fast timing estimates exist now without vendor tools: PyRTL (--syn_tool pyrtl) and a sky130 standard-cell timing model (--part sky130, src/DEVICE_MODELS.py).
